src/internal/operations/export.go

```diff
@@ -0,0 +1,423 @@
+		"service", op.Selectors.Service)
+
+	defer func() {
+		op.bus.Event(
+			ctx,
+			events.ExportEnd,
+			map[string]any{
+				events.BackupID:      op.BackupID,
+				events.DataRetrieved: op.Results.BytesRead,
+				events.Duration:      op.Results.CompletedAt.Sub(op.Results.StartedAt),
+				events.EndTime:       dttm.Format(op.Results.CompletedAt),
+				events.ItemsRead:     op.Results.ItemsRead,
+				events.ItemsWritten:  op.Results.ItemsWritten,
+				events.Resources:     op.Results.ResourceOwners,
+				events.ExportID:      opStats.exportID,
+				events.Service:       op.Selectors.Service.String(),
+				events.StartTime:     dttm.Format(op.Results.StartedAt),
+				events.Status:        op.Status.String(),
+			})
+	}()
+
+	// -----
+	// Execution
+	// -----
+
+	expCollections, err := op.do(ctx, &opStats, sstore, start)
+	if err != nil {
+		// No return here!  We continue down to persistResults, even in case of failure.
+		logger.CtxErr(ctx, err).Error("running export")
+
+		if errors.Is(err, kopia.ErrNoRestorePath) {
+			op.Errors.Fail(clues.New("empty backup or unknown path provided"))
+		}
+
+		op.Errors.Fail(clues.Wrap(err, "running export"))
+	}
+
+	finalizeErrorHandling(ctx, op.Options, op.Errors, "running export")
+	LogFaultErrors(ctx, op.Errors.Errors(), "running export")
+
+	// -----
+	// Persistence
+	// -----
+
+	err = op.persistResults(ctx, start, &opStats)
+	if err != nil {
+		op.Errors.Fail(clues.Wrap(err, "persisting export results"))
+		return nil, op.Errors.Failure()
+	}
+
+	logger.Ctx(ctx).Infow("completed export", "results", op.Results)
+
+	return expCollections, nil
+}
+
+type zipExportCol struct {
+	reader io.ReadCloser
+}
+
+func (z zipExportCol) GetBasePath() string {
+	return ""
+}
+
+func (z zipExportCol) GetItems(ctx context.Context) <-chan export.Item {
+	rc := make(chan export.Item, 1)
+	defer close(rc)
+
+	rc <- export.Item{
+		Data: export.ItemData{
+			Name: "export.zip",
+			Body: z.reader,
+		},
+	}
+
+	return rc
+}
+
+// zipExportCollection takes a list of export collections and zips
+// them into a single collection.
+func zipExportCollection(
+	ctx context.Context,
+	expCollections []export.Collection,
+) (export.Collection, error) {
+	if len(expCollections) == 0 {
+		return nil, clues.New("no export collections provided")
+	}
+
+	reader, writer := io.Pipe()
+	wr := zip.NewWriter(writer)
+
+	go func() {
+		defer writer.Close()
+		defer wr.Close()
+
+		for _, ec := range expCollections {
+			folder := ec.GetBasePath()
+			items := ec.GetItems(ctx)
+
+			for item := range items {
+				err := item.Error
+				if err != nil {
+					writer.CloseWithError(clues.Wrap(err, "getting export item").With("id", item.ID))
+					return
+				}
+
+				name := item.Data.Name
+
+				f, err := wr.Create(folder + "/" + name)
+				if err != nil {
+					writer.CloseWithError(clues.Wrap(err, "creating zip entry").With("name", name).With("id", item.ID))
+					return
+				}
+
+				_, err = io.Copy(f, item.Data.Body)
```

Comment:
we may want to allocate a temp buffer at the start of this function and use `CopyBuffer` instead. The [godoc](https://pkg.go.dev/io#CopyBuffer) for `CopyBuffer` makes it sound like just `Copy` may cause many allocations for buffers. If we go that route always providing a buffer will at least give us some performance guarantees without having to worry about the underlying implementations of the readers/writers (i.e. if they implement other interfaces)